From 9a2f98dbc7d0cfc82985ceaca4ce84ab783089f2 Mon Sep 17 00:00:00 2001
From: TheGamerKing561 <eyadtamer2121@gmail.com>
Date: Sun, 22 Mar 2026 05:00:24 +0200
Subject: [PATCH] Use bms/capacity for percentage, and keep Health HAL for charging status

In my experience, /sys/class/power_supply/bms/capacity gives the correct battery capacity for our device.

This patch forces TWRP to use BMS for battery percentage, and the health HAL for battery charging indicator.
---
 twrp.cpp | 21 ++++++++++++++++++++-
 1 file changed, 20 insertions(+), 1 deletion(-)

diff --git a/twrp.cpp b/twrp.cpp
index c9f597a..9254069 100644
--- a/twrp.cpp
+++ b/twrp.cpp
@@ -461,7 +461,26 @@ int main(int argc, char **argv) {
 			} else {
 				charging = ' ';
 			}
-			lastVal = battery_info.capacity;
+
+			// use battery capacity from bms/capacity instead of health HAL
+			FILE* cap = fopen("/sys/class/power_supply/bms/capacity", "rt");
+			if (cap) {
+				char cap_s[4];
+				fgets(cap_s, sizeof(cap_s), cap);
+				fclose(cap);
+
+				lastVal = atoi(cap_s);
+				if (lastVal > 100) lastVal = 101;
+				if (lastVal < 0) lastVal = 0;
+			} else {
+				// fallback to HAL if file fails and print to log
+				static bool logged_failure = false;
+				if (!logged_failure) {
+					LOGERR("Failed to open /sys/class/power_supply/bms/capacity, using HAL\n");
+					logged_failure = true;
+				}
+				lastVal = battery_info.capacity;
+			}
 #endif
 			// Format the value based on the background updates
 			value = std::to_string(lastVal) + "%" + charging;
-- 
2.53.0.windows.1

